corehq/project_limits/models.py

```diff
@@ -97,3 +101,52 @@ class Meta:
     def __str__(self):
         domain = f"[{self.domain}] " if self.domain else ""
         return f"{domain}{self.key}: {self.limit}"
+
+    def save(self, *args, **kwargs):
+        super().save(*args, **kwargs)
+        self._get_global_limit.clear(self.key)
+        cache.delete(self._specific_limit_cache_key(self.key, self.domain))
+
+    def delete(self, *args, **kwargs):
+        super().delete(*args, **kwargs)
+        self._get_global_limit.clear(self.key)
+        cache.delete(self._specific_limit_cache_key(self.key, self.domain))
+
+    @classmethod
+    def for_key(cls, key, domain=''):
+        """
+        Return the value associated with the given key, prioritizing the domain specific entry over the general one
+        """
+        domain_limit = cls._get_domain_specific_limit(key, domain) if domain else None
+        if domain_limit is not None:
+            return domain_limit
+        return cls._get_global_limit(key)
+
+    @staticmethod
+    @quickcache(['key'], timeout=7 * 24 * 60 * 60, skip_arg=lambda *args, **kwargs: settings.UNIT_TESTING)
```

Comment:
Forgot to mention this earlier, but test code mixed with production code is not ideal. Would it be possible to use a patch to disable this quickcache for the tests that need to disable it?

Might be able to do this with a package- or module-scoped pytest fixture similar to

https://github.com/dimagi/commcare-hq/blob/a95068b2a6cbe30d6496526e53c198d7a873ed8f/corehq/apps/experiments/tests/test_experiment.py#L194-L200

It will probably be a more complicated patch than those, depending on how surgical you need to be to disable quickcache.

Reply:
> Would it be possible to use a patch to disable this quickcache for the tests that need to disable it?

What I don't like about this is that I imagine tests (like the test_device_rate_limiter) that use SystemLimit under the hood would then need to know to disable cache specifically for that SystemLimit method. To your point, maybe just hitting the database is better since it avoids caching entirely, but I think if we do use cache for this method, the best way to reduce friction when writing tests for classes that rely on SystemLimit is to ensure the cache is disabled by default.

Reply:
On the flip side, I would like a way to _enable_ the cache so that I could test that it actually clears how I expect it to in `test_system_limit`, but maybe that is overkill.

Reply:
We offlined, but came to the agreement that using a local cache would ensure that even if this class sees high usage and thus high load, hitting a local cache would be most ideal to avoid network requests. We also agreed that removing the complexity of clearing the cache with the tradeoff of waiting up to 5 minutes for a limit change to take effect was worth it. 

d6a6defcfc4e24b3de37e6c9602af10519fbff37